src/main/java/org/tikv/common/region/RegionManager.java

```diff
@@ -228,12 +228,25 @@ public Pair<TiRegion, TiStore> getRegionStorePairByKey(
 
     TiStore store = null;
     if (storeType == TiStoreType.TiKV) {
-      Peer peer = region.getCurrentReplica();
-      store = getStoreById(peer.getStoreId(), backOffer);
+      // check from the first replica in case it recovers
+      List<Peer> replicaList = region.getReplicaList();
+      for (int i = 0; i < replicaList.size(); i++) {
+        Peer peer = replicaList.get(i);
+        store = getStoreById(peer.getStoreId(), backOffer);
```

Comment:
Is it possible to make replicas not balance ? As we always try the replica 0 at first.

How about start from `getCurrentReplica()`, then if the current replica is not available, get next one by `getNextReplica` and retry.

Reply:
There are two problems if we getNextReplica.
1. code will trap into endless loop when there is no available store.
2. assume we set `follower,leader`, the replica list will begin with the followers then the leaders.  If we getNextReplica, we may miss the recover follower and request the leader

Reply:
> 1. code will trap into endless loop when there is no available store.

It is not difficult to be solved by loop no more than `replicaList.size()` times.

> 2. assume we set `follower,leader`, the replica list will begin with the followers then the leaders.  If we getNextReplica, we may miss the recover follower and request the leader

Why we should not "miss the recover follower and request the leader" ? If we don't want to read on leader, we can set the replica selector as "followers only". Otherwise, if "leader and follower" is used, it should be OK to read on leader.

> Besides, the original code always select the first one. also has the issue of balance

I also feel strange that why `getNextReplica` is unused, and except which `replicaIdx` is changed nowhere. Any idea @iosmanthus ?

Reply:
The customer wants a "fallback" strategy. They want to read followers first to isolate dump traffic. But if all the followers are down, fallback to leaders as a cover.

So if "follower and leader" is specified, it means use followers first, and then leaders. The order matters.

Reply:
Is stale read enabled ? If not, when all followers are down, the leader is also unavailable (neither readable nor writable as well, see https://www.pingcap.com/blog/lease-read/)

Reply:
Ahh, I think we take some terms wrong, it's not follower, it's learner, who will not attend voting. And yes, this is for stale read. The normal TiKV nodes will still take leader and follower and handle online traffic. And there will be extra learner nodes dedicated for offline traffic. When learners are down, goes for the normal nodes.